### PR TITLE
If write op returns ssl_error_want_read, don't write until SSL connec…

### PR DESCRIPTION
**Pull request by xiewei20082008**
…tion is established.

### CODE REVIEW

**Comment by xiewei20082008** [comment]
Fix issue https://github.com/actor-framework/actor-framework/issues/1060.

File: libcaf_io/src/io/network/stream.cpp [hunk 1]
```diff
@@ -143,6 +144,13 @@ bool stream::handle_read_result(rw_state read_result, size_t rb) {
     case rw_state::indeterminate:
       return false;
     case rw_state::success:
+      // if it is the first rw_state::success after SSL_connect,
+      // the ssl connection is established. We recover previous pending write OP.
+      if(wr_op_backoff_) {
+        backend().add(operation::write, fd(), this);
+        wr_op_backoff_ = false;
+      }
+    case rw_state::ssl_error_want_read:
```
Comment (Neverlord) [inline libcaf_io/src/io/network/stream.cpp:153]: Logically, I think `read_some` cannot report `ssl_error_want_read`. That would strike me as an error.

**Review by Neverlord — CHANGES_REQUESTED** [review]
Thanks a lot for digging into this! The patch looks reasonable. I found some minor issues with formatting and I think seeing `want_read` as the result of a read operation would be an error. I have pushed my fixes directly to an integration brach (https://github.com/actor-framework/actor-framework/tree/integration/1276). Could you run that branch to double-check it's still working in your environment?

**Comment by xiewei20082008** [comment]
I have tested the branch with a few lines of log added. https://github.com/xiewei20082008/actor-framework/tree/integration/1276

- Your branch now works, no CPU usage and spinning issue.
- Actually `read_some` will report `ssl_error_want_read` once for every SSL handshake, but as the rb is 0, so it should not be an issue.

```
2021-06-16T12:26:37.895 caf TRACE actor0 6068 do_some ENTRY len = 32 debug_name = "read_some"
2021-06-16T12:26:37.895 caf DEBUG actor0 6068 do_some read_some : connecting
2021-06-16T12:26:37.901 caf DEBUG actor0 6068 operator () SSL_ERROR_WANT_READ reported
2021-06-16T12:26:37.901 caf TRACE actor0 6068 operator () EXIT
2021-06-16T12:26:37.901 caf TRACE actor0 6068 operator () EXIT
2021-06-16T12:26:37.901 caf TRACE actor0 6068 operator () EXIT
2021-06-16T12:26:37.901 caf ERROR actor0 6068 handle_read_result handle_read_result encountered rw_state::want_read, rb= 0
2021-06-16T12:26:37.901 caf TRACE actor0 6068 operator () EXIT
2021-06-16T12:26:37.901 caf TRACE actor0 6068 operator () EXIT
2021-06-16T12:26:37.901 caf TRACE actor0 6068 handle_internal_events ENTRY num-events = 0
```

**Comment by xiewei20082008** [comment]
Thank you for taking the time to review the code. If you need any input from me, just let me know.

**Comment by Neverlord** [comment]
>     * Your branch now works, no CPU usage and spinning issue.

👍

>     * Actually `read_some` will report `ssl_error_want_read` once for every SSL handshake, but as the rb is 0, so it should not be an issue.

Yeah, I should've checked the docs: https://www.openssl.org/docs/man1.1.1/man3/SSL_get_error.html. OpenSSL can return the error for `read` to indicate that it didn't receive enough data to make progress. I probably switch back to your version for the switch-case.

> Thank you for taking the time to review the code. If you need any input from me, just let me know.

Thanks for testing the branch! I probably just revert the handling of wants-read when reading and merge, but if something comes up I'll let you know. 🙂

File: libcaf_io/src/io/network/stream.cpp (hunk 1 above)
Comment (Neverlord) [reply libcaf_io/src/io/network/stream.cpp:153]: Nevermind, read can return this error: https://www.openssl.org/docs/man1.1.1/man3/SSL_read.html.

**Comment by Neverlord** [comment]
@xiewei20082008 FYI, I've merged your changes. GitHub only doesn't recognize the merge since you've pushed to the PR (8c5dc9b) after I've opened my integration branch.

Thanks again for the bugfix! 🙏

**Comment by xiewei20082008** [comment]
> @xiewei20082008 FYI, I've merged your changes. GitHub only doesn't recognize the merge since you've pushed to the PR ([8c5dc9b](https://github.com/actor-framework/actor-framework/commit/8c5dc9b73edc46cec4796be6f13029450d71112e)) after I've opened my integration branch.
> 
> Thanks again for the bugfix!

Cheers!

**Pull request closed without merging**